### PR TITLE
Refactor cuda kernels to custom_op + autograd registration

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary by CodeRabbit

* **New Features**
  * Optional CuPy custom-op support with a runtime toggle and stable object registry; new CUDA-backed single- and multi-step neuron kernels offering autograd-compatible fast paths.

* **Bug Fixes**
  * Improved profiler robustness for CUDA tensor accounting.
  * Fixed convolution FLOP accounting edge case.

* **Refactor**
  * Reorganized neuron-kernel package surface to consolidate exports and simplify kernel entrypoints.

* **Tests**
  * Added CUDA/CuPy integration, torch.compile, and migration tests.

### CODE REVIEW

**Comment by AllenYolk** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by AllenYolk** [comment]
@coderabbitai please review this PR for correctness and maintainability.\n@gemini-code-assist[bot] please review this PR for code quality and potential issues.

**Comment by AllenYolk** [comment]
@coderabbitai review\n@gemini-code-assist[bot] please review latest commit 70eeda92.

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit f8ae34c0.

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit e152990b.

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit 9dde8d20.

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit ae1cc0c3.

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit 5d1d9866.

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit.

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit.

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit.

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit.

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit.

**Comment by AllenYolk** [comment]
@gemini-code-assist[bot] please review latest commit f9eb0d4c for unresolved issues only.
@coderabbitai review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit for unresolved issues only.

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist[bot] please review latest commit for unresolved issues only.

**Comment by AllenYolk** [comment]
@gemini-code-assist[bot] please review latest commit 58f6ca76 for unresolved issues only.
@coderabbitai review

**Comment by AllenYolk** [comment]
@gemini-code-assist[bot] please review latest commit for unresolved issues only.

**Comment by AllenYolk** [comment]
@gemini-code-assist[bot] please review latest commit fd4afadb for unresolved issues only.

**Comment by AllenYolk** [comment]
@gemini-code-assist[bot] please run a fresh review on latest commit fd4afadb and only report unresolved issues.

**Comment by AllenYolk** [comment]
@gemini-code-assist[bot] please review latest commit for unresolved issues only.

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review the latest commit 5e266c58 only.\n\nValidation completed on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nBoth passed (full regression: 447 passed, 2 skipped).

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit e61efdf3 only.\n\nAddressed your note about overly restrictive  gating in spike convolution (now checks only when conv grads are needed).\n\nValidation on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults: targeted 29 passed; full regression 447 passed, 2 skipped.

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit e61efdf3 only.\n\nAddressed your note about restrictive cpp_wrapper gating in spike convolution: now it raises only when convolution gradients (grad_spike/grad_weight) are required, and allows bias-only grad path without cpp wrapper.\n\nValidation on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults: targeted 29 passed; full regression 447 passed, 2 skipped.

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit 1dce9133 only.\n\nAddressed your surrogate API inconsistency concern by adding a unified resolver in cuda neuron kernel common path: accepts both callable cuda_code and callable cuda_codes, then consistently passes resolved callable into custom-op replay and legacy apply paths.\n\nValidation on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults: targeted 29 passed; full regression 447 passed, 2 skipped.

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit 48d82e0a only.\n\nAddressed meta/fake tensor compatibility in python_object_registry_key: tensor identity now avoids unconditional data_ptr() and falls back safely for meta or unsupported tensor types.\n\nValidation on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults: targeted 29 passed; full regression 447 passed, 2 skipped.

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit 8e652188 only.\n\nAddressed your conv-context normalization concern:\n- Added robust normalization for stride/padding/dilation based on input rank.\n- Preserved string padding semantics ("same"/"valid").\n- For CUDA + string padding, spike_conv{1,2,3}d now directly falls back to torch F.conv* path.\n\nValidation on g2:\n- pytest -q test/activation_based/test_pr_new_utilities.py::test_spike_conv1d_cuda_string_padding_fallback\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults: 1 passed; 29 passed; full regression 447 passed, 2 skipped.

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit 993bc36f only.\n\nAddressed PLIF backward efficiency note:\n- Added recovery of full v_v_seq as a view from v_seq storage in setup_context.\n- Cached recovered v_v_seq in ctx and removed backward-time torch.cat reconstruction.\n- Fallback to one-time setup-time concat only if view recovery is impossible.\n\nValidation on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults: targeted 29 passed; full regression 447 passed, 2 skipped.

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit dcb4421d only.\n\nAddressed registry key sorting robustness:\n- Replaced direct dict key tuple sorting with stable key tokens  for heterogeneous keys.\n- Applied this to both general dict normalization and object state normalization in python_object_registry_key.\n\nValidation on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults: targeted 29 passed; full regression 447 passed, 2 skipped.

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit dcb4421d only.\n\nAddressed registry key sorting robustness:\n- Replaced direct dict key tuple sorting with stable key tokens of key class module, key class qualname, and key repr for heterogeneous keys.\n- Applied this to both general dict normalization and object state normalization in python_object_registry_key.\n\nValidation on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults: targeted 29 passed; full regression 447 passed, 2 skipped.

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit d7f7ae17 only.\n\nAddressed two points:\n1) int32 launch metadata safety: kept int32 (matches CUDA kernel signatures) and added explicit range guard with clear OverflowError in auto_cuda neuron_kernel common path.\n2) documented spike convolution backward scope as non-transposed-only with fixed zero output_padding in wrapper helper.\n\nValidation on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults: targeted 29 passed; full regression 447 passed, 2 skipped.

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit 8f9444c0 only.\n\nAddressed the eager performance regressions from replay-based custom-op backwards:\n- ss_if_step / ss_lif_step now route to custom-op path only during graph capture (compile/tracing); eager path uses direct ATGF apply.\n- multistep_qif_ptt / multistep_izhikevich_ptt / multistep_eif_ptt now also gate custom-op usage to graph capture only; eager path stays on direct MultiStep*PTT.apply.\n\nValidation on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults: targeted 29 passed; full regression 447 passed, 2 skipped.

**Comment by AllenYolk** [comment]
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit 9ad185ba only for unresolved issues.\n\nAddressed your latest review points:\n1) Added fp16 odd-neuron padding/unpadding in single-step CuPy IF/LIF paths.\n2) Prevented capture-context leak by stashing only when any forward input tensor requires grad; sentinel token used otherwise.\n3) Standardized ss kernel launch metadata numel to int32 cupy scalars.\n\nValidation on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults:\n- targeted: 29 passed\n- full regression: 444 passed, 2 skipped\n

**Comment by AllenYolk** [comment]
@gemini-code-assist Please review latest commit 1b227796 only for unresolved issues.\n\nAddressed your latest summary items with minimal, concrete fixes:\n1) Added ss kernel object-id caching (WeakKeyDictionary) to avoid repeated register/key generation every step in ss IF/LIF custom-op path.\n2) Standardized ss scalar int conversion to int32 in common scalar_to_cupy path.\n3) Added explicit power-of-two threads guard for PLIF backward reduction kernel (configure.cuda_threads).\n\nValidation on g2:\n- pytest -q test/activation_based/test_cuda_custom_op_migration.py test/activation_based/test_cupy_backend_comprehensive.py\n- pytest -q test/activation_based --ignore=test/activation_based/test_op_counter_neuromc_runtime.py\n\nResults:\n- targeted: 29 passed\n- full regression: 444 passed, 2 skipped\n
